Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [3]:
import re

hd=[]
with open('/Users/ruaykd/Desktop/untitled folder/horse_names.txt', mode = 'r') as fl:
    for ln in fl.readlines():
        try:
            hd.append(ln[re.search(r'\d+:', ln).end()+1:-1].strip())
        except:
            pass
hd

['surgery?',
 'Age',
 'Hospital Number',
 'rectal temperature',
 'pulse',
 'respiratory rate',
 'temperature of extremities',
 'peripheral pulse',
 'mucous membranes',
 'capillary refill time',
 "pain - a subjective judgement of the horse's pain level",
 'peristalsis',
 'abdominal distension',
 'nasogastric tube',
 'nasogastric reflux',
 'nasogastric reflux PH',
 'rectal examination - feces',
 'abdomen',
 'packed cell volume',
 'total protein',
 'abdominocentesis appearance',
 'abdomcentesis total protein',
 'outcome',
 'surgical lesion?',
 'type of lesion',
 'cp_data']

In [4]:
import pandas as pd

fl = '/Users/ruaykd/Desktop/untitled folder/horse_data.csv'

df = pd.read_csv(fl, header = None, names=hd, usecols = range(len(hd)), na_values = '?')

df.info()
df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 26 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   surgery?                                                 299 non-null    float64
 1   Age                                                      300 non-null    int64  
 2   Hospital Number                                          300 non-null    int64  
 3   rectal temperature                                       240 non-null    float64
 4   pulse                                                    276 non-null    float64
 5   respiratory rate                                         242 non-null    float64
 6   temperature of extremities                               244 non-null    float64
 7   peripheral pulse                                         231 non-null    float64
 8   mucous membranes              

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0         2.0    1           530101                38.5   66.0   
1         1.0    1           534817                39.2   88.0   
2         2.0    1           530334                38.3   40.0   
3         1.0    9          5290409                39.1  164.0   
4         2.0    1           530255                37.3  104.0   
..        ...  ...              ...                 ...    ...   
295       1.0    1           533886                 NaN  120.0   
296       2.0    1           527702                37.2   72.0   
297       1.0    1           529386                37.5   72.0   
298       1.0    1           530612                36.5  100.0   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
0                28.0                         3.0               3.0   
1                20.0                         NaN               NaN   
2                24.0                         1.0               1.0   
3                84.0                         4.0               1.0   
4                35.0                         NaN               NaN   
..                ...                         ...               ...   
295              70.0                         4.0               NaN   
296              24.0                         3.0               2.0   
297              30.0                         4.0               3.0   
298              24.0                         3.0               3.0   
299              20.0                         NaN               NaN   

     mucous membranes  capillary refill time  ...  rectal examination - feces  \
0                 NaN                    2.0  ...                         3.0   
1                 4.0                    1.0  ...                         4.0   
2                 3.0                    1.0  ...                         1.0   
3                 6.0                    2.0  ...                         3.0   
4                 6.0                    2.0  ...                         NaN   
..                ...                    ...  ...                         ...   
295               4.0                    2.0  ...                         NaN   
296               4.0                    2.0  ...                         4.0   
297               4.0                    1.0  ...                         3.0   
298               3.0                    1.0  ...                         4.0   
299               NaN                    NaN  ...                         4.0   

     abdomen  packed cell volume  total protein  abdominocentesis appearance  \
0        5.0                45.0            8.4                          NaN   
1        2.0                50.0           85.0                          2.0   
2        1.0                33.0            6.7                          NaN   
3        NaN                48.0            7.2                          3.0   
4        NaN                74.0            7.4                          NaN   
..       ...                 ...            ...                          ...   
295      5.0                55.0           65.0                          NaN   
296      4.0                44.0            NaN                          3.0   
297      5.0                60.0            6.8                          NaN   
298      4.0                50.0            6.0                          3.0   
299      1.0                36.0           62.0                          1.0   

     abdomcentesis total protein  outcome  surgical lesion?  type of lesion  \
0                            NaN      2.0                 2           11300   
1                            2.0      3.0                 2            2208   
2                            NaN      1.0                 2               0   
3                            5.3      2.0                 1            2208   
4                            N

In [5]:
# анализируем отдельные параметры

#v = 'surgery?'
#v = 'Age'
#v = 'Hospital Number'
#v = 'rectal temperature'
#v = 'pulse'
#v = 'respiratory rate'
#v = 'temperature of extremities'
v = 'capillary refill time'

print(v)
print('количество    =', df[v].count())
print('минимум  0%   =', df[v].min())
print('        25%   =', df[v].quantile(0.25))
print('медиана 50%   =', df[v].median())
print('        75%   =', df[v].quantile(0.75))
print('максимум 100% =', df[v].max())
print('размах        =', df[v].max() - df[v].min())
print('ср.значение   =', df[v].mean())
print('мода          =', df[v].mode()[0])
print('ско           =', df[v].std())
print('дисперсия     =', df[v].var())
print('уникумы       =', df[v].unique())
print('в количестве  =', len(list(df[v].unique())))
print('и количество их повторений:')
print(' - по частоте')
print(df[v].value_counts(dropna = False))
print(' - по значению параметра')
print(df[v].value_counts(dropna = True).sort_index())   

capillary refill time
количество    = 268
минимум  0%   = 1.0
        25%   = 1.0
медиана 50%   = 1.0
        75%   = 2.0
максимум 100% = 3.0
размах        = 2.0
ср.значение   = 1.3059701492537314
мода          = 1.0
ско           = 0.47762853540401734
дисперсия     = 0.22812901783218667
уникумы       = [ 2.  1. nan  3.]
в количестве  = 4
и количество их повторений:
 - по частоте
1.0    188
2.0     78
NaN     32
3.0      2
Name: capillary refill time, dtype: int64
 - по значению параметра
1.0    188
2.0     78
3.0      2
Name: capillary refill time, dtype: int64


# Результаты анализа

surgery?:
- одно пропущенное значение;
- 60% животных оперировалось;

Age:
- пропущенных значений нет;
- из описания: может принимать только 2 значения: 1 (взрослая лошадь) или 2 (молодая лошадь), по факту также 2 значения, но 1 и 9;
- если предположить, что 9 - это ошибка ввода, то, вероятно, что 9 можно заменить на 2; 9 считать выбросом нельзя;
- подавляющая статистика собиралась по взрослым особям (Age = 1): 276 против 24;

Hospital Number -  ID лошади:
- пропущенных значений нет;
- данные составлены на основании обследования 284 животных;
- некоторые животные лечились дважды;

rectal temperature - ректальная температура:
- пропущено 60 значений, что составляет 20% наблюдений;
- из описания: нормальная температура 37.8, а это первая квартиль, соответственно в 75% случаев температура была повышенной;
- о повышенной температуре говорят и среднее значение, и мода с медианой;
- явные выбросы отсутствуют, а как минимальные, так и максимальные значения могут свидетельстовать об определенных особенностях состояния здоровья;

pulse - пульс:
- пропущено 24 значения, что составляет 8% наблюдений;
- если верить описанию (30-40 норма, редко м.б. ниже), то у подавляющего числа животных (92%) был учащенный пульс;
- наиболее часто наблюдалось значение 48, что на 20% выше верхней границы нормы;
- среднее значение пульса у пациентов (71.9) почти в 2 раза выше нормы;
- коэффициент вариации ско/ср.знач. * 100 = 40% говорит об очень сильном разбросе значений;
- на мой взгляд явных выбросов нет, особенно учитывая комментарии в описании; в данном случае средние и отклонения от них ничего не дают (или распределение полимодально, но уж несимметрично относительно средних), а подбирая критерии, например, четырех сигм (а не трех), можно показать, что выбросов нет;

respiratory rate - частота дыхания:
- пропущено 58 значений, что составляет почти 20% наблюдений;
- по описанию 8 - 10 это норма, тогда только у 2.5% животных она была нормальной, а в 97.5% учащенной;
- среднее значение 30.4, что в 3 раза выше нормы;
- коэффициент вариации ско/ср.знач. * 100 = 58% говорит об очень сильном разбросе значений;
- не наблюдается какого-либо группирования значений в одном районе;
- мысли по поводу выбросов аналогичны тем, что были по пульсу; м.б. если бы количество пропусков было меньше, то картина была бы более четкой; сейчас +3сигмы дают 80, что 8-10 раз больше нормы, но это вполне может объясняться проявлением болезненного состояния животного;

temperature of extremities - температура конечностей:
- пропущено 56 значений, что составляет 18.7% наблюдений;
- нормальная температура наблюдалась в 32% случаев, самая частая - прохладная - 44.6%, холодная - в 11% и теплая - в 12.4%
- выбросов нет;

capillary refill time - время наполнения капилляров:
- пропущено 32 значения, что составляет 11% наблюдений;
- из описания: может принимать только 2 значения: 1 и 2; здесь присутствуют 3: 1, 2 и 3. Возможно, 3 - ошибка, возможно, что данные брались из разных источников, соответственно, были разные шкалы и т.п
- но т.к. значение 3 имеется только в 2-х случаях, то можно попробовать его игнорировать;
- в большинстве случаев (70%) наблюдалась недостаточная циркуляция крови;
- выбросов нет;


По поводу выбросов и пропущенных значений:
- "явные" (грубые) выбросы можно отсеивать, применяя различные критерии, но, во-первых, почти всегда можно найти "правильный" критерий, который покажет, что это не выброс, а, во-вторых, при такой оценке нужно использовать знания из предметной области, а не только из статистики и теории вероятостей. К тому же большинство критериев ориентируется на нормальные или близкие к нему распределения, что в рассмотренных случаях не очень соблюдалось. 
- замена пропущенных значений чем-либо - очень тонкая вещь, которая без предметных знаний просто опасна. Можно оценить значимость влияния параметров на что-то изучаемое, и, если у каких-то мало значащих параметров много пропусков, то их можно исключить из рассмотрения. Любая замена приводит к искажению результатов.

In [6]:
# смотрим, как влияет количество не пустых значений на количество остающихся строк
for i in range(1, len(df.axes[1])+1):
    print('если в строке по меньшей мере', i, 'не пустое значение (', len(df.axes[1]) - i,'пропусков), то остается записей', len(df.dropna(thresh = i).axes[0]))

если в строке по меньшей мере 1 не пустое значение ( 25 пропусков), то остается записей 300
если в строке по меньшей мере 2 не пустое значение ( 24 пропусков), то остается записей 300
если в строке по меньшей мере 3 не пустое значение ( 23 пропусков), то остается записей 300
если в строке по меньшей мере 4 не пустое значение ( 22 пропусков), то остается записей 300
если в строке по меньшей мере 5 не пустое значение ( 21 пропусков), то остается записей 300
если в строке по меньшей мере 6 не пустое значение ( 20 пропусков), то остается записей 300
если в строке по меньшей мере 7 не пустое значение ( 19 пропусков), то остается записей 300
если в строке по меньшей мере 8 не пустое значение ( 18 пропусков), то остается записей 296
если в строке по меньшей мере 9 не пустое значение ( 17 пропусков), то остается записей 296
если в строке по меньшей мере 10 не пустое значение ( 16 пропусков), то остается записей 292
если в строке по меньшей мере 11 не пустое значение ( 15 пропусков), то остаетс

In [7]:
# смотрим, как влияет количество не пустых значений на количество остающихся строк на отобранных параметрах
df2 = df[['surgery?', 'Age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities','capillary refill time']]
#df2
df2.info()
for i in range(1, len(df2.axes[1])+1):
    print('если в строке по меньшей мере', i, 'не пустое значение (', len(df2.axes[1]) - i,'пропусков), то остается записей', len(df2.dropna(thresh = i).axes[0]))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   Hospital Number             300 non-null    int64  
 3   rectal temperature          240 non-null    float64
 4   pulse                       276 non-null    float64
 5   respiratory rate            242 non-null    float64
 6   temperature of extremities  244 non-null    float64
 7   capillary refill time       268 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.9 KB
если в строке по меньшей мере 1 не пустое значение ( 7 пропусков), то остается записей 300
если в строке по меньшей мере 2 не пустое значение ( 6 пропусков), то остается записей 300
если в строке по меньшей мере 3 не пустое значение ( 5 пропусков), то остается записей 300
ес

В полном наборе данных полностью заполнены только 6 строк ...
Максимальное количество пропусков (18 штук) наблюдается в 4 строках.

На моем наборе данных полностью заполнены 164 строки (чуть больше половины). Из параметров хуже всего с "respiratory rate" (частотой дыхания) и "temperature of extremities" (температурой конечностей) - заполнено порядка 80% строк.

In [8]:
# Заменим в ранговых параметрах ('surgery?', 'temperature of extremities', 'capillary refill time') отсутствующие значения
# на константу 0 - отсутствие данных (там нет такого ранга)
# а 'rectal temperature', 'pulse' и 'respiratory rate' на медиану по группам ранговых переменных
df2['surgery?'].fillna(value = 0.0, inplace = True)
df2['temperature of extremities'].fillna(value = 0.0, inplace = True)
df2['capillary refill time'].fillna(value = 0.0, inplace = True)

#print(df2.groupby(['surgery?', 'temperature of extremities', 'capillary refill time'])['rectal temperature'].median())

df2['rectal temperature'].fillna(value = df2.groupby(['surgery?', 'temperature of extremities', 'capillary refill time'])['rectal temperature'].transform('median').round(1), inplace=True)
df2['pulse'].fillna(value = df2.groupby(['surgery?', 'temperature of extremities', 'capillary refill time'])['pulse'].transform('median').round(1), inplace=True)
df2['respiratory rate'].fillna(value = df2.groupby(['surgery?', 'temperature of extremities', 'capillary refill time'])['respiratory rate'].transform('median').round(1), inplace=True)

# а незамененных - на их моды
df2['rectal temperature'].fillna(value = df2['rectal temperature'].mode()[0], inplace=True)
df2['pulse'].fillna(value = df2['pulse'].mode()[0], inplace=True)
df2['respiratory rate'].fillna(value = df2['respiratory rate'].mode()[0], inplace=True)

df2.info()
df2.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   Hospital Number             300 non-null    int64  
 3   rectal temperature          300 non-null    float64
 4   pulse                       300 non-null    float64
 5   respiratory rate            300 non-null    float64
 6   temperature of extremities  300 non-null    float64
 7   capillary refill time       300 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.9 KB


/Users/ruaykd/opt/anaconda3/lib/python3.8/site-packages/pandas/core/series.py:4463: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


surgery?         Age  Hospital Number  rectal temperature  \
count  300.000000  300.000000     3.000000e+02          300.000000   
mean     1.393333    1.640000     1.085889e+06           38.157667   
std      0.496094    2.173972     1.529801e+06            0.661943   
min      0.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.875000   
50%      1.000000    1.000000     5.303055e+05           38.100000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory rate  temperature of extremities  \
count  300.000000        300.000000                  300.000000   
mean    71.253333         29.503333                    1.910000   
std     27.893775         16.234794                    1.314404   
min     30.000000          8.000000                    0.000000   
25%     48.000000         20.000000                    1.000000   
50%     62.000000         24.000000                    2.000000   
75%     88.000000         35.250000                    3.000000   
max    184.000000         96.000000                    4.000000   

       capillary refill time  
count             300.000000  
mean                1.166667  
std                 0.605622  
min                 0.000000  
25%                 1.000000  
50%                 1.000000  
75%                 2.000000  
max                 3.000000